src/connection/mod.rs

```diff
@@ -114,49 +126,129 @@ impl<C: FirebirdClient> Connection<C> {
         res
     }
 
-    //cleans up statement cache and releases the database handle
+    // Cleans up statement cache and releases the database handle
     fn cleanup_and_detach(&mut self) -> Result<(), FbError> {
         StmtCache::close_all(self);
 
+        // Drop the default transaction
+        if let Some(mut tr) = self.def_tr.take() {
+            tr.rollback(self).ok();
+        }
+
         self.cli.detach_database(&mut self.handle)?;
 
         Ok(())
     }
 
     /// Run a closure with a transaction, if the closure returns an error
-    /// the transaction will rollback, else it will be committed
+    /// and the default transaction is not active, the transaction will rollback, else it will be committed
     pub fn with_transaction<T, F>(&mut self, closure: F) -> Result<T, FbError>
     where
         F: FnOnce(&mut Transaction<C>) -> Result<T, FbError>,
     {
-        let mut tr = Transaction::new(self)?;
+        let in_transaction = self.in_transaction;
+
+        let mut tr = if let Some(tr) = self.def_tr.take() {
+            tr.into_transaction(self)
+        } else {
+            Transaction::new(self)?
+        };
 
         let res = closure(&mut tr);
 
-        if res.is_ok() {
-            tr.commit_retaining()?;
+        if !in_transaction {
+            if res.is_ok() {
+                tr.commit_retaining()?;
+            } else {
+                tr.rollback_retaining()?;
+            }
+        }
+
+        let tr = TransactionData::from_transaction(tr);
+
+        if let Some(mut tr) = self.def_tr.replace(tr) {
+            // Should never happen, but just to be sure
+            tr.rollback(self).ok();
+        }
+
+        res
+    }
+
+    /// Run a closure with the default transaction, no rollback or commit will be automatically performed
+    /// after the closure returns. The next call to this function will use the same transaction
+    /// if it was not closed with `commit_retaining` or `rollback_retaining`
+    fn use_transaction<T, F>(&mut self, closure: F) -> Result<T, FbError>
```

Comment:
```suggestion
    pub fn use_transaction<T, F>(&mut self, closure: F) -> Result<T, FbError>
```

should it be `pub`?

Reply:
I think it shouldn't be. To me, this method is for internal utilization. 

Do you have any example where we need this method in the outside?

Reply:
not really, just wondered why it had a doc comment in this case and wondered if pub was omitted by mistake